### PR TITLE
Add ticketmaster API 

### PR DESCRIPTION
**Pull request by eylulguleryuz**
Added the ticketmaster API on top of the pull request #11. Couldn't test it out with LLM studio and all the other APIs because of some errors but as a stand-alone it works and returns results. I also tried to handle the cases with the date. When the user inputs something like "today", "tomorrow", or "24-11" , they are converted into proper format and sent to the API.

### CODE REVIEW

File: backend/app/api/events.py [hunk 1]
```diff
@@ -54,10 +56,78 @@ def query_eventbrite(city: str, date: str, keywords: List[str]) -> List[Event]:
     
     return events
 
+def get_formatted_date(date_entity): 
+    current_date = datetime.now() 
+    # Handle relative dates like "today" or "tomorrow"
+    if date_entity:
```
Comment (bartekleon) [inline backend/app/api/events.py:62]: how about:
```py
if not date_entity:
   return None, None

rest of the logic without additional indent
```

File: backend/app/api/events.py [hunk 2]
```diff
@@ -54,10 +56,78 @@ def query_eventbrite(city: str, date: str, keywords: List[str]) -> List[Event]:
...
+        date_entity = date_entity.lower()
+        if "today" in date_entity:
+            return current_date.replace(hour=0, minute=0, second=0, microsecond=0).isoformat(), current_date.replace(hour=23, minute=59, second=59, microsecond=0).isoformat()
```
Comment (bartekleon) [inline backend/app/api/events.py:65]: the today and tomorrow could be basically written as one statement:

```py
if "today" in date_entity or "tomorrow" in date_entity:
   delta = timedelta(days=1) if "tomorrow" else timedelta(days=0)
   start = current_date.replace(hour=0, minute=0, second=0, microsecond=0) + delta 
   end = current_date.replace(hour=23, minute=59, second=59, microsecond=0) + delta 
   return start.isoformat(), end.isoformat()
```

Comment (bartekleon) [reply backend/app/api/events.py:65]: or make a function (so we handle all 4 cases)
```py
def get_start_end_date(date):
  start_date = parsed_date.replace(hour=0, minute=0, second=0, microsecond=0)
  end_date = parsed_date.replace(hour=23, minute=59, second=59, microsecond=0)
  return start_date.isoformat(), end_date.isoformat()
```

and then we have:

```py
if "today" in date_entity:
   return get_start_end_date(datetime.now())
elif "tomorrow" in date_entity:
   return get_start_end_date(datetime.now() + timedelta(days=1))
else:
      # Handle YYYY-MM-DD format
      try:
          return get_start_end_date(datetime.strptime(date_entity, "%Y-%m-%d"))
      except ValueError:
          # Handle specific formats like "23-11" (dd-mm)
          try:
             return get_start_end_date(datetime.strptime(date_entity, "%d-%m").replace(year=current_date.year))
          except ValueError:
              return None, None
```

File: backend/app/api/events.py [hunk 3]
```diff
@@ -54,10 +56,78 @@ def query_eventbrite(city: str, date: str, keywords: List[str]) -> List[Event]:
...
+                # Handle specific formats like "23-11" (dd-mm)
+                try:
+                    parsed_date = datetime.strptime(date_entity, "%d-%m").replace(year=current_date.year)
+                    start_date = parsed_date.replace(hour=0, minute=0, second=0, microsecond=0)
+                    end_date = parsed_date.replace(hour=23, minute=59, second=59, microsecond=0)
+                    return start_date.isoformat(), end_date.isoformat()
+                except ValueError:
+                    return None, None
+                
+def query_ticketmaster(city: str, date: str, keywords: List[str]) -> List[Event]:
```
Comment (bartekleon) [inline backend/app/api/events.py:87]: `List` can be `list`

File: backend/app/api/events.py [hunk 4]
```diff
@@ -54,10 +56,78 @@ def query_eventbrite(city: str, date: str, keywords: List[str]) -> List[Event]:
...
+            id=event_data.get("id"),
+            name=event_data.get("name"),
+            location=city,  # Event location set to the input city
+            url=event_data.get("url"),
+            date=event_data["dates"]["start"]["localDate"],
+            category = event_data.get("classifications", [])[0].get("segment", {}).get("name", "Unknown Category")
+        )
+        events.append(event)
+    if events == []:
+        return "No events found."
```
Comment (bartekleon) [inline backend/app/api/events.py:124]: Everywhere you return a list, yet here you return `string`? Smells like a bug

File: backend/app/query.py [hunk 5]
```diff
@@ -62,24 +62,28 @@ def run_query(query: QueryRequest):
     if pinned_events:
         return QueryResponse(id=query.session_id, message=f"Prioritized event: {pinned_events}")
 
+    event_results = []
```
Comment (bartekleon) [inline backend/app/query.py:65]: Why assign it? it's literally assigned 2 lines below?

File: backend/app/query.py [hunk 6]
```diff
@@ -62,24 +62,28 @@ def run_query(query: QueryRequest):
...
     # Step 3: Fetch events from OpenTripMap
     event_results = query_opentripmap(OpenTripMapModel(placename=city, kinds=keywords))
 
-    # Step 4: Fetch weather from OpenWeatherMap
+    # Step 4: Fetch dynamic events from Ticketmaster and append them to OpenTripMap results
+    event_results.append(events.query_ticketmaster(city, date, keywords))
```
Comment (bartekleon) [inline backend/app/query.py:70]: I would assume `event_results` is `list[Event]` type. Your result is also `list[Event]`

You probably wanted to use `extend` and not `append`.
In your case you would have something like:
________________                  added by you - v
```[Event, Event, ..., Event, list[Event]]```

And we probably want `[Event, Event, Event, ..., Event, Event]`

File: backend/app/api/events.py [hunk 7]
```diff
@@ -54,10 +56,78 @@ def query_eventbrite(city: str, date: str, keywords: List[str]) -> List[Event]:
...
+    filtered_keywords = [kw for kw in keywords if kw.lower() != city.lower() and kw.lower() != date.lower()]
+    # Set the day start and end, handle cases like "tomorrow", "today"
+    date, end_date = get_formatted_date(date)
+    print(f"Querying Ticketmaster: city={city}, date={date}, endDate={end_date}, keywords={filtered_keywords}")
+
+    params = {
+        "keyword": " ".join(filtered_keywords),
+        "city": f"{city}",
+        "sort": "date,asc",
+        "startDateTime": date + "+00:00",
```
Comment (bartekleon) [inline backend/app/api/events.py:99]: What if `date` and `end_date` are `None`?

File: backend/app/api/events.py [hunk 8]
```diff
@@ -54,10 +56,78 @@ def query_eventbrite(city: str, date: str, keywords: List[str]) -> List[Event]:
...
+        "endDateTime": end_date + "+00:00",
+        "apikey": TICKETMASTER_API_KEY,
+        "size": 5 #page size of the response
+        }
+    response = requests.get(TICKETMASTER_API_URL, params=params)
```
Comment (bartekleon) [inline backend/app/api/events.py:104]: What happens if we can't reach the server? Please check

**Review by bartekleon — CHANGES_REQUESTED** [review]
3 potential bugs and one nitpicking :)

File: backend/app/api/events.py (hunk 7 above)
Comment (eylulguleryuz) [reply backend/app/api/events.py:99]: they wouldn't be cuz the chat screens warns the user if no date is entered
```
 city = info.get('city')
    date = info.get('date')
    keywords = info.get('keywords')
    if city is None or date is None:
        CITY_NOT_FOUND = "You seem to have not provided the city or date correctly. Please double check it"
```

File: backend/app/query.py (hunk 6 above)
Comment (eylulguleryuz) [reply backend/app/query.py:70]: fixed

File: backend/app/query.py (hunk 5 above)
Comment (eylulguleryuz) [reply backend/app/query.py:65]: fixed

File: backend/app/api/events.py (hunk 4 above)
Comment (eylulguleryuz) [reply backend/app/api/events.py:124]: fixed

File: backend/app/api/events.py (hunk 3 above)
Comment (eylulguleryuz) [reply backend/app/api/events.py:87]: fixed

File: backend/app/api/events.py (hunk 2 above)
Comment (eylulguleryuz) [reply backend/app/api/events.py:65]: fixed

File: backend/app/api/events.py (hunk 1 above)
Comment (eylulguleryuz) [reply backend/app/api/events.py:62]: fixed

File: backend/app/api/events.py (hunk 8 above)
Comment (eylulguleryuz) [reply backend/app/api/events.py:104]: umm anything other than success should fall here?
```
    response = requests.get(TICKETMASTER_API_URL, params=params)
    if response.status_code != 200:
        print(f"TicketMaster API Error: {response.status_code} {response.text}")
        return []
```
but maybe you mean sth else?

**Pull request closed without merging**